kinto_portier/portier.py

```diff
@@ -0,0 +1,121 @@
+import json
+import re
+from base64 import urlsafe_b64decode
+from datetime import timedelta
+from six.moves.urllib.request import urlopen
+
+import jwt
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives.asymmetric import rsa
+
+
+def b64dec(string):
+    """Decode unpadded URL-safe Base64 strings.
+    Base64 values in JWTs and JWKs have their padding '=' characters stripped
+    during serialization. Before decoding, we must re-append padding characters
+    so that the encoded value's final length is evenly divisible by 4.
+    """
+    padding = '=' * ((4 - len(string) % 4) % 4)
+    return urlsafe_b64decode(string + padding)
+
+
+def discover_keys(broker, cache):
+    """Discover and return a Broker's public keys.
+    Returns a dict mapping from Key ID strings to Public Key instances.
+    Portier brokers implement the `OpenID Connect Discovery`_ specification.
+    This function follows that specification to discover the broker's current
+    cryptographic public keys:
+    1. Fetch the Discovery Document from ``/.well-known/openid-configuration``.
+    2. Parse it as JSON and read the ``jwks_uri`` property.
+    3. Fetch the URL referenced by ``jwks_uri`` to retrieve a `JWK Set`_.
+    4. Parse the JWK Set as JSON and extract keys from the ``keys`` property.
+    Portier currently only supports keys with the ``RS256`` algorithm type.
+    .. _OpenID Connect Discovery:
+        https://openid.net/specs/openid-connect-discovery-1_0.html
+    .. _JWK Set: https://tools.ietf.org/html/rfc7517#section-5
+    """
+    # Check the cache
+    cache_key = 'jwks:' + broker
+    raw_jwks = cache.get(cache_key)
+    if not raw_jwks:
+        # Fetch Discovery Document
+        res = urlopen(''.join((broker, '/.well-known/openid-configuration')))
+        discovery = json.loads(res.read().decode('utf-8'))
+        if 'jwks_uri' not in discovery:
+            raise RuntimeError('No jwks_uri in discovery document')
```

Comment:
That's imported code from https://github.com/portier/demo-rp/
So I have no strong opinion of moving from one to the other.